styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -321,36 +321,28 @@ public void updateNextNaturalTrigger(WorkflowId workflowId, TriggerInstantSpec t
   }
 
   public Map<WorkflowId, Workflow> workflows(Set<WorkflowId> workflowIds) {
-    final Callable<Map<WorkflowId, Workflow>> callable =
-        () -> Lists
-            .partition(ImmutableList.copyOf(workflowIds), MAX_NUMBER_OF_ENTITIES_IN_ONE_BATCH)
-            .parallelStream()
-            .map(this::getBatchOfWorkflows)
-            .flatMap(Collection::stream)
-            .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
-    return forkJoinPool.submit(callable).join();
-  }
-
-  private Set<Map.Entry<WorkflowId, Workflow>> getBatchOfWorkflows(final List<WorkflowId> batch) {
-    final Set<Key> keys = batch.stream()
-        .map(workflowId -> workflowKey(datastore.newKeyFactory(), workflowId))
-        .collect(toSet());
-    final Iterator<Entity> result = datastore.get(keys);
+    final Iterable<List<WorkflowId>> batches = Iterables.partition(workflowIds, MAX_NUMBER_OF_ENTITIES_IN_ONE_BATCH);
+    return StreamSupport.stream(batches.spliterator(), false)
+        .map(batch -> forkJoinPool.submit(() -> this.getBatchOfWorkflows(batch)))
+        .collect(toList())
```

Comment:
Is this needed?

Reply:
Doesn't seem to be needed. At least to make it compile.

Reply:
So much difference in performance! Without `collect` and `stream` it is much slower.

Reply:
OK I get it. Java `collect` is magic.

Reply:
Without `collect` it will run just one task at a time =)

This behavior is certainly not obvious, should add a comment about the need for collect + stream.

Reply:
I will add it.